File: p3.js
// ## Problem3  Rotate Array by K Places
// (https://leetcode.com/problems/rotate-array/)

// time: O(n)
// space: O(1)

const rotate = function(nums, k) {
    if(!nums || nums.length < 2) return;
    let n = nums.length - 1;
    if(k > nums.length) k = k % nums.length;
    reverse(nums, 0 ,n);
    reverse(nums, 0, k-1);
    reverse(nums, k, n);
};
function reverse(nums, l, r) {
    while(l < r) {
        let temp = nums[l];
        nums[l] = nums[r];
        nums[r] = temp;
        l++;
        r--;
    }
}

File: p3.java
// Time Complexity :O(n)
// Space Complexity :O(1)
// Did this code successfully run on Leetcode :
// Any problem you faced while coding this :


// Your code here along with comments explaining your approach


class Solution {
    public void rotate(int[] nums, int k) {
        k = k%nums.length;
        if(k==0) return;
        //Reverse the whole array
        for(int i=0; i<nums.length/2; i++){
            int temp = nums[i];
            nums[i] = nums[nums.length-1-i];
            nums[nums.length-1-i] = temp;
        }

       
        //Reverse first k elements
        for(int i=0; i<k/2; i++){
            int temp = nums[i];
            nums[i] = nums[k-1-i];
            nums[k-1-i] = temp;
        }

        //Reverse remaining elements
        for(int i=k; i<(nums.length+k)/2; i++){
            int temp = nums[i];
            nums[i] = nums[nums.length-1-i+k];
            nums[nums.length-1-i+k] = temp;
        }

         
    }
}